petab_select/candidate_space.py

```diff
@@ -529,6 +602,628 @@ def iterate():
+            and not [
+                critical_set
+                for critical_set in self.critical_parameter_sets
+                if len(critical_set) > 1
+            ]
+            and not self.swap_parameter_sets
+        ):
+            if self.number_of_reattempts:
+                self.jump_to_most_distant()
+                return
+            else:
+                raise StopIteration(
+                    "The next chosen method is Method.LATERAL, but there are no crit or swap parameters provided. Terminating"
+                )
+        if previous_method == Method.LATERAL:
+            self.swap_done_successfully = False
+        self.update_method(method=next_method)
+
+    def update_method(self, method: Method):
+        """Update self.method to the method."""
+        if method not in [Method.FORWARD, Method.LATERAL, Method.BACKWARD]:
+            raise NotImplementedError(
+                f'FAMoS direction must be either `Method.FORWARD`, `Method.BACKWARD` or `Method.LATERAL`, not {method}. \
+                Check if the method_scheme scheme provided is correct.'
+            )
+
+        self.method = method
+
+    def switch_inner_candidate_space(self, history):
+        """Switch self.inner_candidate_space to the candidate space of
+        the current self.method."""
+
+        self.inner_candidate_space = self.inner_candidate_spaces[self.method]
+        # reset the next inner candidate space with the current history of all checked models
+        self.inner_candidate_space.reset(
+            predecessor_model=self.predecessor_model, exclusions=history.keys
+        )
+        # FIXME Check that problem.calibrated_models is excluded.
+
+    def jump_to_most_distant(self):
+        """Jump to most distant model with respect to the history of all
+        calibrated models."""
+
+        new_init_model = self.get_most_distant()
+
+        logging.info("JUMPING: ", new_init_model.parameters)
+        # breakpoint()
+
+        # if model not appropriate make it so by adding the first
+        # critical parameter from each critical parameter set
+        if not self.check_critical(new_init_model):
+            for critical_set in self.critical_parameter_sets:
+                new_init_model.parameters[critical_set[0]] = ESTIMATE
+
+        self.update_method(self.initial_method)
+
+        self.number_of_reattempts -= 1
+        self.jumped_to_most_distant = True
+
+        self.predecessor_model = new_init_model
+        self.best_model_of_current_run = new_init_model
+        if self.summary_tsv is not None:
+            with open(self.summary_tsv, 'a') as f:
+                writer = csv.writer(f, delimiter='\t')
+                writer.writerow(
+                    [
+                        "Jumped to the most distant model.",
+                        *([''] * 5),
+                    ]
+                )
+
+    # TODO Fix for non-famos model subspaces. FAMOS easy beacuse of only 0;ESTIMATE
```

Comment:
check

Reply:
Yes, this is one of the big TODOs, to make get_most_distant more general, for more complicated model_spaces

Reply:
The comment says `0;ESTIMATE`, but is the `0` important or can the `0` be changed to any fixed value that represents "off", e.g. if a parameter is a multiplicative constant, then "off" would be `1`.

If `0` is important then it would be nice to change, since we already have examples where `1` is used to turn off a parameter. However, I guess that `0` is not important, since we already got the Blasi model working, where `1` represents "off".